### PR TITLE
Secret providers

### PR DESCRIPTION
**Pull request by GunaKKIBM**

### CODE REVIEW

File: provider/iam/token_exchange.go [hunk 1]
```diff
@@ -117,13 +125,15 @@ func (tes *tokenExchangeService) ExchangeIAMAPIKeyForIMSToken(iamAPIKey string,
...
 
 // ExchangeIAMAPIKeyForAccessToken ...
-func (tes *tokenExchangeService) ExchangeIAMAPIKeyForAccessToken(iamAPIKey string, logger *zap.Logger) (*AccessToken, error) {
-	r := tes.newTokenExchangeRequest(logger)
-
-	r.request.Field("grant_type", "urn:ibm:params:oauth:grant-type:apikey")
-	r.request.Field("apikey", iamAPIKey)
-
-	return r.exchangeForAccessToken()
+func (tes *tokenExchangeService) ExchangeIAMAPIKeyForAccessToken(logger *zap.Logger) (*AccessToken, error) {
```
Comment (sameshai) [inline provider/iam/token_exchange.go:128]: Any plan to change function name ? or due to many providers using this are we keeping same name ?

Comment (GunaKKIBM) [reply provider/iam/token_exchange.go]: Since, this is used by file vpc also, won't be changing the name currently.

**Review by sameshai — APPROVED** [review]
LGTM

**Review by skorati — APPROVED** [review]
LGTM

**Review by sameshai — COMMENTED** [review]
LGTM

**Review by prankulmahajan — APPROVED** [review]
lgtm

File: provider/auth/ims_token.go [hunk 2]
```diff
@@ -83,17 +82,6 @@ func (ccf *ContextCredentialsFactory) ForIAMAccessToken(apiKey string, logger *z
...
 
-// UpdateAPIKey ...
-func (ccf *ContextCredentialsFactory) UpdateAPIKey(apiKey string, logger *zap.Logger) error {
-	logger.Info("Updating api key")
-	if ccf.TokenExchangeService == nil {
-		logger.Error("Failed to update api key in context credentials")
-		return errors.New("failed to update api key")
-	}
-	err := ccf.TokenExchangeService.UpdateAPIKey(apiKey, logger)
-	return err
```
Comment (GunaKKIBM) [inline provider/auth/ims_token.go:94]: Revert these changes for the time being.

Comment (GunaKKIBM) [inline provider/auth/ims_token.go:94]: Revert these changes for the time being.

File: provider/iam/token_exchange.go [hunk 3]
```diff
@@ -30,12 +30,15 @@ import (
 	"github.com/IBM-Cloud/ibm-cloud-cli-sdk/common/rest"
 	"github.com/IBM/ibmcloud-volume-interface/config"
 	util "github.com/IBM/ibmcloud-volume-interface/lib/utils"
+	"github.com/IBM/secret-common-lib/pkg/secret_provider"
+	sp "github.com/IBM/secret-utils-lib/pkg/secret_provider"
 )
 
 // tokenExchangeService ...
 type tokenExchangeService struct {
```
Comment (GunaKKIBM) [inline provider/iam/token_exchange.go:38]: Removing Auth Config Later - Add it as an action item.

File: provider/iam/token_exchange.go [hunk 4]
```diff
@@ -30,12 +30,15 @@ import (
...
-	authConfig *AuthConfiguration
-	httpClient *http.Client
+	authConfig     *AuthConfiguration
+	httpClient     *http.Client
+	secretprovider sp.SecretProviderInterface
```
Comment (GunaKKIBM) [inline provider/iam/token_exchange.go:41]: Consider using this as a part of the session interface - Action item.

File: provider/iam/token_exchange.go [hunk 5]
```diff
@@ -118,12 +126,14 @@ func (tes *tokenExchangeService) ExchangeIAMAPIKeyForIMSToken(iamAPIKey string,
...
 // ExchangeIAMAPIKeyForAccessToken ...
 func (tes *tokenExchangeService) ExchangeIAMAPIKeyForAccessToken(iamAPIKey string, logger *zap.Logger) (*AccessToken, error) {
-	r := tes.newTokenExchangeRequest(logger)
-
-	r.request.Field("grant_type", "urn:ibm:params:oauth:grant-type:apikey")
-	r.request.Field("apikey", iamAPIKey)
-
-	return r.exchangeForAccessToken()
+	logger.Info("Fetching using secret provider")
+	token, _, err := tes.secretprovider.GetDefaultIAMToken(false)
```
Comment (GunaKKIBM) [inline provider/iam/token_exchange.go:130]: Check the required token lifetime for CSI driver
